kratos/includes/data_communicator.h

```diff
@@ -0,0 +1,360 @@
+    {
+        return rLocalValue;
+    }
+
+    virtual int Min(const int rLocalValue, const int Root) const
+    {
+        return rLocalValue;
+    }
+
+    virtual double Min(const double rLocalValue, const int Root) const
+    {
+        return rLocalValue;
+    }
+
+    virtual array_1d<double,3> Min(const array_1d<double,3>& rLocalValue, const int Root) const
+    {
+        return rLocalValue;
+    }
+
+    virtual int Max(const int rLocalValue, const int Root) const
+    {
+        return rLocalValue;
+    }
+
+    virtual double Max(const double rLocalValue, const int Root) const
+    {
+        return rLocalValue;
+    }
+
+    virtual array_1d<double,3> Max(const array_1d<double,3>& rLocalValue, const int Root) const
+    {
+        return rLocalValue;
+    }
+
+    // Allreduce operations
+
+    virtual int SumAll(const int rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual double SumAll(const double rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual array_1d<double,3> SumAll(const array_1d<double,3>& rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual int MinAll(const int rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual double MinAll(const double rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual array_1d<double,3> MinAll(const array_1d<double,3>& rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual int MaxAll(const int rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual double MaxAll(const double rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual array_1d<double,3> MaxAll(const array_1d<double,3>& rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    // Scan operations
+
+    virtual int ScanSum(const int rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    virtual double ScanSum(const double rLocalValue) const
+    {
+        return rLocalValue;
+    }
+
+    // Sendrecv operations
+
+    virtual void SendRecv(
+        const std::vector<int>& rSendValues, const unsigned int SendDestination,
+        std::vector<int>& rRecvValues, const unsigned int RecvSource) const
+    {}
+
+    virtual void SendRecv(
+        const std::vector<double>& rSendValues, const unsigned int SendDestination,
+        std::vector<double>& rRecvValues, const unsigned int RecvSource) const
+    {}
+
+    // Broadcast
+
+    virtual void Broadcast(
+        std::vector<int>& rBuffer,
+        const unsigned int SourceRank) const
+    {}
+
+    virtual void Broadcast(
+        std::vector<double>& rBuffer,
+        const unsigned int SourceRank) const
+    {}
+
+    // Scatter operations
+
+    virtual void Scatter(
+        const std::vector<int>& rSendValues,
+        std::vector<int>& rRecvValues,
+        const unsigned int SourceRank) const
+    {}
```

Comment:
Note that `DataCommunicator` functions that work on std::vectors are not really equivalent to their MPI counterparts (on mpi, the "output" buffer is filled in the end, on serial runs it is left uninitialized). I am open for discussion regarding this, but for now my idea is to check if it is a distributed run (`DataCommunicator::IsDistributed() == true`) and only create/fill the buffer std::vectors if so.

I would like to do something more robust, but this would require actually copying the buffers (send->recv) for no good reason in serial runs, plus it opens a can of worms regarding what to do if the buffers are not of the right size (should I throw an error if a buffer that is not really needed has the wrong size?)

Reply:
+1 for rGlobalValues = rLocalValues or @pooyan-dadvand's above suggestion. The function should perform its task as advertised and the caller should only call it if this is what the caller wants imo.